### PR TITLE
Picam Software

### PR DESCRIPTION
**Pull request by cooperqb**
Repurposed version of the vent/valve code which reflects the functionality of the picam project.

---
This change is [Reviewable](https://reviewable.io/reviews/waterloo-rocketry/cansw_picam/1)

### CODE REVIEW

**Review by BluCodeGH — CHANGES_REQUESTED** [review]
Looking good, has this been tested on hardware?

> Reviewed 19 of 33 files at r1, 7 of 16 files at r2, 9 of 9 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/cansw_picam/1#-:-My-urkFA21p-MKrWjpz:b-igd5wm)* status: all files reviewed, 4 unresolved discussions (waiting on @cooperqb, @ZTeper, and @zx100x100)

___
*[cam.h, line 2 at r3](https://reviewable.io/reviews/waterloo-rocketry/cansw_picam/1#-My-sPQr04F_a_wSfW32:-My-sPQr04F_a_wSfW33:b-u0sk0f) ([raw file](https://github.com/waterloo-rocketry/cansw_picam/blob/ed8128b02a4b738ef143ae4821eee16c83ee1e3b/cam.h#L2)):*
> ```objc
> #ifndef CAM_H
> #define	CAM_H
> ```

Sneaky tab here
___
*[cam.c, line 81 at r3](https://reviewable.io/reviews/waterloo-rocketry/cansw_picam/1#-My-siPw6ViyXM3Y8o44:-My-siPw6ViyXM3Y8o45:bjlis8u) ([raw file](https://github.com/waterloo-rocketry/cansw_picam/blob/ed8128b02a4b738ef143ae4821eee16c83ee1e3b/cam.c#L81)):*
> ```C
>     build_cam_stat_msg(millis(), curr_state, req_state, &stat_msg); // canlib-->can_common.h 
>     txb_enqueue(&stat_msg);
> }
> ```

Since this code was designed for a mechanical system with 4 possible states it doesn't map well to the picam. It would probably be cleaner to replace this all with a single `get_cam_state()` function that `cam_send_status()` calls and returns either CAM_ON or CAM_OFF based on, since there is no way we can have an in-between / invalid state here.
___
*[main.c, line 26 at r3](https://reviewable.io/reviews/waterloo-rocketry/cansw_picam/1#-My-tFtC1kpLK7IfSfEV:-My-tFtC1kpLK7IfSfEW:bwhsxxu) ([raw file](https://github.com/waterloo-rocketry/cansw_picam/blob/ed8128b02a4b738ef143ae4821eee16c83ee1e3b/main.c#L26)):*
> ```C
> // Follows CAM_STATE in message_types.h
> // SHOULD ONLY BE MODIFIED IN ISR
> static enum CAM_STATE requested_cam_state = CAM_ON;
> ```

Do we want the default state to be off or on? `cam.c` implies off (see `cam_init()`).
___
*[README.md, line 1 at r3](https://reviewable.io/reviews/waterloo-rocketry/cansw_picam/1#-My-h3I_-nMHbbundm5y:-My-h3I_-nMHbbundm5z:boclea5) ([raw file](https://github.com/waterloo-rocketry/cansw_picam/blob/ed8128b02a4b738ef143ae4821eee16c83ee1e3b/README.md#L1)):*
> ```Markdown
> ### Vent Board Control Software
> ```

README needs an update (or well needs to be written lmao)

**Review by BluCodeGH — CHANGES_REQUESTED** [review]
> Reviewed 10 of 10 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/cansw_picam/1)* status: all files reviewed, 8 unresolved discussions (waiting on @cooperqb, @ZTeper, and @zx100x100)

___
*[cam.c, line 51 at r4](https://reviewable.io/reviews/waterloo-rocketry/cansw_picam/1#-Myjo81y62xfk4jhjGxy:-Myjo81y62xfk4jhjGxz:b-z3tj24) ([raw file](https://github.com/waterloo-rocketry/cansw_picam/blob/88b3f1890e0b5df0b1a211dd4de079e2ba0014d1/cam.c#L51)):*
> ```C
>     // This is the same as the injector valve status.
>     enum ACTUATOR_STATE curr_state;
>     enum ACTUATOR_ID actuator_id = PICAM;
> ```

The actuator id should be defined globally in a .h file somewhere. I'd consider bringing back `board.h` to define board-specific constants like the CAN board id (which I don't think you have defined anywhere atm) and this.
___
*[cam.c, line 52 at r4](https://reviewable.io/reviews/waterloo-rocketry/cansw_picam/1#-Myjnu0H2M0GOofYN31q:-Myjnu0H2M0GOofYN31r:b-anci55) ([raw file](https://github.com/waterloo-rocketry/cansw_picam/blob/88b3f1890e0b5df0b1a211dd4de079e2ba0014d1/cam.c#L52)):*
> ```C
>     enum ACTUATOR_STATE curr_state;
>     enum ACTUATOR_ID actuator_id = PICAM;
>     curr_state = get_cam_state();
> ```

We should use an explicit ternary operator here (`get_cam_state() ? ACTUATOR_OPEN : ACTUATOR_CLOSE`) rather than relying on the ordering of the enum, also helps make what's happening more clear.
___
*[main.c, line 110 at r4](https://reviewable.io/reviews/waterloo-rocketry/cansw_picam/1#-Myjord_69t-TMJUiGxu:-Myjorda7b8_kixzHEpE:bk9mbzp) ([raw file](https://github.com/waterloo-rocketry/cansw_picam/blob/88b3f1890e0b5df0b1a211dd4de079e2ba0014d1/main.c#L110)):*
> ```C
>             if((last_millis%500) == 0)
>                 cam_send_status(requested_cam_state);
> ```

This if statement only runs every `MAX_LOOP_TIME_DIFF_ms`, easiest to just send the status every time (I'd do it after the line where we `send_status_ok()`). Also you can't guarantee that we will hit every value of `millis()` sequentially (if the loop takes more than 1ms to run some will be skipped) so its very unlikely the modulo will ever be exactly zero.
___
*[main.c, line 135 at r4](https://reviewable.io/reviews/waterloo-rocketry/cansw_picam/1#-MyjpNt23RcdjruAZIJz:-MyjpNt3EToGWf656e4l:b2sm14) ([raw file](https://github.com/waterloo-rocketry/cansw_picam/blob/88b3f1890e0b5df0b1a211dd4de079e2ba0014d1/main.c#L135)):*
> ```C
> static void can_msg_handler(const can_msg_t *msg) {
>     uint16_t msg_type = get_message_type(msg);
>     static enum ACTUATOR_ID actuator_id = PICAM;
> ```

move to .h as mentioned.
___
*[main.c, line 157 at r4](https://reviewable.io/reviews/waterloo-rocketry/cansw_picam/1#-MyjpWnJ0BACPyN19gu9:-MyjpWnK57Cc7nqZYGgq:bhofehq) ([raw file](https://github.com/waterloo-rocketry/cansw_picam/blob/88b3f1890e0b5df0b1a211dd4de079e2ba0014d1/main.c#L157)):*
> ```C
>         //case MSG_ACTUATOR_STATUS:
>         //    cam_send_status(requested_cam_state);
>         //    break;
> ```

Delete.
___
*[main.c, line 166 at r4](https://reviewable.io/reviews/waterloo-rocketry/cansw_picam/1#-MyjpeV6-EiXhbpxgM3h:-MyjpeV6-EiXhbpxgM3i:bjzz0cq) ([raw file](https://github.com/waterloo-rocketry/cansw_picam/blob/88b3f1890e0b5df0b1a211dd4de079e2ba0014d1/main.c#L166)):*
> ```C
>                     cam_on();
>                 else if(requested_cam_state == ACTUATOR_CLOSED)
>                     cam_off();
> ```

We already handle turning on and off the camera in the main loop, probably best to not duplicate code since this isn't a super time critical action.
___
*[README.md, line 12 at r4](https://reviewable.io/reviews/waterloo-rocketry/cansw_picam/1#-Myjpnix0VHBpVJ-H7BT:-Myjpnix0VHBpVJ-H7BU:b-feu38h) ([raw file](https://github.com/waterloo-rocketry/cansw_picam/blob/88b3f1890e0b5df0b1a211dd4de079e2ba0014d1/README.md#L12)):*
> ```Markdown
> * Load the code onto the board using MPLAB.
> 
> When updating code, add Jack Christensen (aka BluCodeGH), Zachary Teper, Cooper Brunke
> ```

I find lines like this go out of date pretty quickly, from what I've seen people requesting based on past PRs / who else is involved in the project has worked well.

File: cam.c
Comment (ZTeper) [inline cam.c]: This should go in cam_init()

**Review by ZTeper — APPROVED** [review]
I approve of this PR, but please wait for the code to be tested on hardware before merging into `master`

**Review by Artem-Sotnikov — COMMENTED** [review]
Also looked at this, looks good. Should be good to merge once code has been tested.

> Reviewed 7 of 7 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/cansw_picam/1#-:-Mz5OdI41BQFb--JN_mL:bndfv3d)* status: all files reviewed, 9 unresolved discussions (waiting on @cooperqb and @zx100x100)

**Review by BluCodeGH — APPROVED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/cansw_picam/1)* status: all files reviewed, 1 unresolved discussion (waiting on @cooperqb and @zx100x100)

**Comment by BluCodeGH** [comment]
I believe this has been tested? Please address comments on reviewable's end (click the reviewable button in the first message in the PR) and merge.

**Review by cooperqb — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/cansw_picam/1)* status: 35 of 36 files reviewed, 1 unresolved discussion (waiting on @Artem-Sotnikov, @ZTeper, and @zx100x100)

___
*[`cam.c` line 81 at r3](https://reviewable.io/reviews/waterloo-rocketry/cansw_picam/1#-My-siPw6ViyXM3Y8o44:-N0wJrdO6i40QbmhvhaO:bb7f2mf) ([raw file](https://github.com/waterloo-rocketry/cansw_picam/blob/ed8128b02a4b738ef143ae4821eee16c83ee1e3b/cam.c#L81)):*
<details><summary><i>Previously, BluCodeGH (BluCode) wrote…</i></summary><blockquote>

Since this code was designed for a mechanical system with 4 possible states it doesn't map well to the picam. It would probably be cleaner to replace this all with a single `get_cam_state()` function that `cam_send_status()` calls and returns either CAM_ON or CAM_OFF based on, since there is no way we can have an in-between / invalid state here.
</blockquote></details>

done
___
*[`cam.c` line 51 at r4](https://reviewable.io/reviews/waterloo-rocketry/cansw_picam/1#-Myjo81y62xfk4jhjGxy:-N0wJibA0wYz_lX8Rvi8:bb7f2mf) ([raw file](https://github.com/waterloo-rocketry/cansw_picam/blob/88b3f1890e0b5df0b1a211dd4de079e2ba0014d1/cam.c#L51)):*
<details><summary><i>Previously, BluCodeGH (BluCode) wrote…</i></summary><blockquote>

The actuator id should be defined globally in a .h file somewhere. I'd consider bringing back `board.h` to define board-specific constants like the CAN board id (which I don't think you have defined anywhere atm) and this.
</blockquote></details>

done
___
*[`cam.c` line 52 at r4](https://reviewable.io/reviews/waterloo-rocketry/cansw_picam/1#-Myjnu0H2M0GOofYN31q:-N0wJkYXA0GakjQNSq8R:bb7f2mf) ([raw file](https://github.com/waterloo-rocketry/cansw_picam/blob/88b3f1890e0b5df0b1a211dd4de079e2ba0014d1/cam.c#L52)):*
<details><summary><i>Previously, BluCodeGH (BluCode) wrote…</i></summary><blockquote>

We should use an explicit ternary operator here (`get_cam_state() ? ACTUATOR_OPEN : ACTUATOR_CLOSE`) rather than relying on the ordering of the enum, also helps make what's happening more clear.
</blockquote></details>

done
___
*[`cam.c` line 21 at r5](https://reviewable.io/reviews/waterloo-rocketry/cansw_picam/1#-MxSAod2lUScr-nt0GZy-r5-21:-N0wJ_0S-OKmKt8XQn7w:b-rnrbll) ([raw file](https://github.com/waterloo-rocketry/cansw_picam/blob/3cc578fadd8e8f6d91eed63dcbcdbfe78e7f86e1/cam.c#L21)):*
<details><summary><i>Previously, ZTeper (Zachary Teper) wrote…</i></summary><blockquote>

This should go in cam_init()
</blockquote></details>

done
___
*[`main.c` line 26 at r3](https://reviewable.io/reviews/waterloo-rocketry/cansw_picam/1#-My-tFtC1kpLK7IfSfEV:-N0wJv8a8MPL4GCElRYH:bb7f2mf) ([raw file](https://github.com/waterloo-rocketry/cansw_picam/blob/ed8128b02a4b738ef143ae4821eee16c83ee1e3b/main.c#L26)):*
<details><summary><i>Previously, BluCodeGH (BluCode) wrote…</i></summary><blockquote>

Do we want the default state to be off or on? `cam.c` implies off (see `cam_init()`).
</blockquote></details>

done
___
*[`main.c` line 110 at r4](https://reviewable.io/reviews/waterloo-rocketry/cansw_picam/1#-Myjord_69t-TMJUiGxu:-N0wK0HI-8rSrl0f4LJz:bb7f2mf) ([raw file](https://github.com/waterloo-rocketry/cansw_picam/blob/88b3f1890e0b5df0b1a211dd4de079e2ba0014d1/main.c#L110)):*
<details><summary><i>Previously, BluCodeGH (BluCode) wrote…</i></summary><blockquote>

This if statement only runs every `MAX_LOOP_TIME_DIFF_ms`, easiest to just send the status every time (I'd do it after the line where we `send_status_ok()`). Also you can't guarantee that we will hit every value of `millis()` sequentially (if the loop takes more than 1ms to run some will be skipped) so its very unlikely the modulo will ever be exactly zero.
</blockquote></details>

done
___
*[`main.c` line 135 at r4](https://reviewable.io/reviews/waterloo-rocketry/cansw_picam/1#-MyjpNt23RcdjruAZIJz:-N0wK1Cc6xFbZGu99GiH:bb7f2mf) ([raw file](https://github.com/waterloo-rocketry/cansw_picam/blob/88b3f1890e0b5df0b1a211dd4de079e2ba0014d1/main.c#L135)):*
<details><summary><i>Previously, BluCodeGH (BluCode) wrote…</i></summary><blockquote>

move to .h as mentioned.
</blockquote></details>

done
___
*[`main.c` line 157 at r4](https://reviewable.io/reviews/waterloo-rocketry/cansw_picam/1#-MyjpWnJ0BACPyN19gu9:-N0wK22DD6y0lzOzGgyI:bb7f2mf) ([raw file](https://github.com/waterloo-rocketry/cansw_picam/blob/88b3f1890e0b5df0b1a211dd4de079e2ba0014d1/main.c#L157)):*
<details><summary><i>Previously, BluCodeGH (BluCode) wrote…</i></summary><blockquote>

Delete.
</blockquote></details>

done
___
*[`main.c` line 166 at r4](https://reviewable.io/reviews/waterloo-rocketry/cansw_picam/1#-MyjpeV6-EiXhbpxgM3h:-N0wK3AB3ke3CIJ9oq-Q:bb7f2mf) ([raw file](https://github.com/waterloo-rocketry/cansw_picam/blob/88b3f1890e0b5df0b1a211dd4de079e2ba0014d1/main.c#L166)):*
<details><summary><i>Previously, BluCodeGH (BluCode) wrote…</i></summary><blockquote>

We already handle turning on and off the camera in the main loop, probably best to not duplicate code since this isn't a super time critical action.
</blockquote></details>

done
___
*[`README.md` line 1 at r3](https://reviewable.io/reviews/waterloo-rocketry/cansw_picam/1#-My-h3I_-nMHbbundm5y:-N0wK4Oj5QMuihEp3zmx:bb7f2mf) ([raw file](https://github.com/waterloo-rocketry/cansw_picam/blob/ed8128b02a4b738ef143ae4821eee16c83ee1e3b/README.md#L1)):*
<details><summary><i>Previously, BluCodeGH (BluCode) wrote…</i></summary><blockquote>

README needs an update (or well needs to be written lmao)
</blockquote></details>

done
___
*[`README.md` line 12 at r4](https://reviewable.io/reviews/waterloo-rocketry/cansw_picam/1#-Myjpnix0VHBpVJ-H7BT:-N0wK5N67NhPzeUz_aEF:bb7f2mf) ([raw file](https://github.com/waterloo-rocketry/cansw_picam/blob/88b3f1890e0b5df0b1a211dd4de079e2ba0014d1/README.md#L12)):*
<details><summary><i>Previously, BluCodeGH (BluCode) wrote…</i></summary><blockquote>

I find lines like this go out of date pretty quickly, from what I've seen people requesting based on past PRs / who else is involved in the project has worked well.
</blockquote></details>

done

**Review by cooperqb — COMMENTED** [review]
> Reviewed 18 of 33 files at r1, 1 of 16 files at r2, 2 of 9 files at r3, 4 of 10 files at r4, 6 of 7 files at r5, 1 of 1 files at r6, all commit messages.
> Dismissed @ZTeper from a discussion.
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/cansw_picam/1)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @zx100x100)

**Pull request merged**
